jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/CredentialManager.kt

```diff
@@ -6,75 +6,151 @@ package software.aws.toolkits.jetbrains.core.credentials
+import java.util.concurrent.ConcurrentHashMap
+import javax.security.auth.login.CredentialNotFoundException
+
+data class CredentialIdentifierChange(
+    val added: List<ToolkitCredentialsIdentifier>,
+    val modified: List<ToolkitCredentialsIdentifier>,
+    val removed: List<ToolkitCredentialsIdentifier>
+)
+
+typealias CredentialsChangeListener = (change: CredentialIdentifierChange) -> Unit
 
 abstract class CredentialManager : SimpleModificationTracker() {
+    private val providerIds = ConcurrentHashMap<String, ToolkitCredentialsIdentifier>()
+    private val awsCredentialProviderCache = ConcurrentHashMap<ToolkitCredentialsIdentifier, ConcurrentHashMap<String, AwsCredentialsProvider>>()
+
+    protected abstract fun factoryMapping(): Map<String, CredentialProviderFactory>
+
     @Throws(CredentialProviderNotFound::class)
-    abstract fun getCredentialProvider(providerId: String): ToolkitCredentialsProvider
+    fun getAwsCredentialProvider(providerId: ToolkitCredentialsIdentifier, region: AwsRegion): ToolkitCredentialsProvider =
+        ToolkitCredentialsProvider(providerId, AwsCredentialProviderProxy(providerId, region))
+
+    fun getCredentialIdentifiers(): List<ToolkitCredentialsIdentifier> = providerIds.values.toList()
+
+    fun getCredentialIdentifierById(id: String): ToolkitCredentialsIdentifier? = providerIds[id]
+
+    // TODO: Convert these to bulk listeners so we only send N messages where N is # of extensions vs # of providers
+    protected fun addProvider(identifier: ToolkitCredentialsIdentifier) {
+        providerIds[identifier.id] = identifier
 
-    abstract fun getCredentialProviders(): List<ToolkitCredentialsProvider>
+        incModificationCount()
+        ApplicationManager.getApplication().messageBus.syncPublisher(CREDENTIALS_CHANGED).providerAdded(identifier)
+    }
+
+    protected fun modifyProvider(identifier: ToolkitCredentialsIdentifier) {
+        awsCredentialProviderCache.remove(identifier)
+
+        incModificationCount()
+        ApplicationManager.getApplication().messageBus.syncPublisher(CREDENTIALS_CHANGED).providerModified(identifier)
+    }
+
+    protected fun removeProvider(identifier: ToolkitCredentialsIdentifier) {
+        providerIds.remove(identifier.id)
+        awsCredentialProviderCache.remove(identifier)
+
+        incModificationCount()
+        ApplicationManager.getApplication().messageBus.syncPublisher(CREDENTIALS_CHANGED).providerRemoved(identifier)
+    }
+
+    /**
+     * Inner class that lazy-ily requests the true AwsCredentialsProvider from the factory when needed. This acts as a middle man that allows for existing
+     * ToolkitCredentialsProvider (like ones passed to existing SDK clients), to keep operating even if the credentials they represent have been updated such
+     * as loading from disk when new values.
+     */
+    private inner class AwsCredentialProviderProxy(private val providerId: ToolkitCredentialsIdentifier, private val region: AwsRegion) :
```

Comment:
nit: this seems like less of a proxy and more of a cache

Reply:
It doesn't actually cache anything. It's a "pointer" that is capable of replacing its underlying aws credential provider. Thats why I call it a proxy.... Better name?

Reply:
sounds good, proxy seems fine